feat: Add item templates feature (#435)

Wrr2216 opened the pull request:
## What type of PR is this?
Feature

## What this PR does / why we need it
Adds Item Templates so users can save reusable templates with default values and custom fields for quick item creation.

## Which issue(s) this PR fixes
Fixes #435

## Special notes for your reviewer

**Backend:**
- `item_template.go` / `template_field.go` – New Ent schemas for templates and their custom fields
- `group.go` – Added templates edge
- `repo_item_templates.go` – Repository with CRUD
- `v1_ctrl_item_templates.go` / `routes.go` – API handlers and routes
- Migrations for SQLite and Postgres

**Frontend:**
- `templates.vue` / `template/[id].vue` – List and detail pages
- `Template/` components – Card, CreateModal, Selector
- `Item/CreateModal.vue` – Integrated template selector
- `item/[id]/index.vue` – "Use as Template" action
- `default.vue` – Nav link
- `templates.ts` – API client

**API Endpoints:**
- `GET/POST /api/v1/templates` – List and create
- `GET/PUT/DELETE /api/v1/templates/{id}` – Read, update, delete
- `POST /api/v1/templates/{id}/create-item` – Create item from template

## Testing
Both builds pass. Manually tested template CRUD, creating items from templates, and the "Use as Template" flow.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * End-to-end item templates: API CRUD, create-item-from-template, templates dashboard, detail page, create/edit modals, card, selector, duplicate/delete, and "save as template" from item pages.
  * Item Create modal can apply templates, pre-filling defaults and restoring last-used template.
  * Template selector UI with searchable popover and template apply/clear flows.

* **Tests**
  * Added comprehensive backend and frontend tests covering template lifecycle and integrations.

* **Documentation**
  * Added API doc comments and new i18n strings for template UI.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tankerkiller125 commented:
I'll look at this deeper tomorrow, hopefully we can get it merged for the next release.

This PR will basically have to make it in before the next release really (and I'll see that it does), as we intend to implement a feature freeze on item/location related things after that release, so that the item/location merge can get completed without dealing with major merge conflicts every week.

tankerkiller125 commented on backend/internal/data/migrations/postgres/20251128120101_drop_item_templates.sql:
```diff
@@ -0,0 +1,3 @@
+-- +goose Up
```
So the only thing here is that the SQLite and Postgres have 3 migrations each for this change.

I know that this may have been needed in a test case/playing with things. However, because this code hasn't yet been published, to production, one migration for each type should work (the final one "recreate" one.

Simply deleting the extra ones, and renaming the "recreate" one should suffice in that area.

Because we can't ever delete migrations (at least not without breaking the migration path from Hay-Kots version to our latest versions) keeping the migrations as clean as possible before they hit the main branch helps maintain things.

Wrr2216 commented:
I swear I had that issue and resolved it, maybe I deleted the fix for that. It is the GUID not passing through in the request properly.

Wrr2216 commented:
@tankerkiller125 I found an issue with the migration files. They had some old rows still there that were not being used. I also found the GUID was sending as a nil string which was a no no to the server. I changed this to a proper GUID styled zeroed out string for each ID and also revised the migrations to remove those old rows. I have tested several different ways of creating the template and they all are passing now. If you wouldn't mind to give it a try and let me know if its good for you.

tonyaellie commented on frontend/components/Item/CreateModal.vue:
```diff
@@ -31,6 +31,35 @@
+        <div class="mb-2 flex items-center justify-between">
+          <h4 class="text-sm font-semibold">Template: {{ templateData.name }}</h4>
+        </div>
+        <div class="space-y-1 text-xs text-muted-foreground">
```
use gap and flex, not space if possible

tonyaellie commented on frontend/components/Item/CreateModal.vue:
```diff
@@ -31,6 +31,35 @@
+        </div>
+        <div class="space-y-1 text-xs text-muted-foreground">
+          <p v-if="templateData.description">{{ templateData.description }}</p>
+          <div class="mt-2 grid grid-cols-2 gap-x-4 gap-y-1">
```
not got firefox installed but should double check if this is fine, i know that in the past firefox has had some issues with grid

tonyaellie commented on frontend/components/Item/CreateModal.vue:
```diff
@@ -31,6 +31,35 @@
+          </div>
+          <div v-if="templateData.fields && templateData.fields.length > 0" class="mt-2">
+            <p class="font-medium">Custom Fields:</p>
+            <ul class="ml-4 list-none space-y-1">
```
again prefer gap over space

tonyaellie reviewed (changes requested):
Templates should also be able to include:
labels, location, item name, item description, maybe model num

Would also be nice if template info on the item creation popup was collapsible

Template should be remembered between creations, similar to location

Don't love this ui for selecting the template, it suggests that the template is part of them item

<img width="988" height="932" alt="Image" src="https://github.com/user-attachments/assets/4470ba29-ab06-42c1-b763-9a933d1a3900" />

I think move it so its next to the barcode stuff, e.g.

<img width="549" height="396" alt="Image" src="https://github.com/user-attachments/assets/e0a3c6a8-c5fe-4a4b-a024-d6d7438fc131" />

Not sure about having templates in the sidebar, but it works for now

The buttons here should be sqaure, there should also be a duplicate option

<img width="798" height="453" alt="Image" src="https://github.com/user-attachments/assets/a959b0b2-fe46-436b-8293-9640b5dddbc1" />




To many places to highlight them all, but all strings should be translated.

Overall I like where this is going but needs some work, for things where the same issue is repeated i did give up on commenting it each time, e.g. space vs gap.

Wrr2216 commented on frontend/pages/item/[id]/index.vue:
```diff
@@ -655,6 +696,10 @@
                 <MdiPlusBoxMultipleOutline />
                 <span class="hidden md:inline">{{ $t("global.duplicate") }}</span>
               </Button>
+              <Button class="w-9 md:w-auto" aria-label="Save as Template" @click="saveAsTemplate">
```
Definitely agree on that. I will look over the rest of your recommendations above. Thanks!

tonyaellie commented on frontend/components/Item/CreateModal.vue:
```diff
@@ -1,8 +1,22 @@
+              />
+            </TooltipTrigger>
+            <TooltipContent>
+              <p>Apply a template</p>
```
Needs translating

tonyaellie commented on frontend/components/Item/CreateModal.vue:
```diff
@@ -31,6 +45,84 @@
+            <p v-if="templateData.description" class="text-foreground/80">{{ templateData.description }}</p>
+            <div class="grid grid-cols-2 gap-x-4 gap-y-1">
+              <div v-if="templateData.defaultName">
+                <span class="font-medium">Name:</span> {{ templateData.defaultName }}
```
Needs translating and the ones below

tonyaellie commented on frontend/components/Item/CreateModal.vue:
```diff
@@ -261,6 +362,94 @@
+    // Load full template details
+    const { data, error } = await api.templates.get(template.id);
+    if (error || !data) {
+      toast.error("Failed to load template details");
```
Needs translating

tonyaellie commented on frontend/pages/templates.vue:
```diff
@@ -0,0 +1,57 @@
+  const { data: templates, refresh } = useAsyncData("templates", async () => {
+    const { data, error } = await api.templates.getAll();
+    if (error) {
+      toast.error("Failed to load templates");
```
Needs translating

tonyaellie commented on frontend/pages/templates.vue:
```diff
@@ -0,0 +1,57 @@
+    </div>
+
+    <div v-else class="flex flex-col items-center justify-center py-12 text-center">
+      <p class="mb-4 text-muted-foreground">No templates yet.</p>
```
Needs translating

tonyaellie commented on frontend/pages/templates.vue:
```diff
@@ -0,0 +1,57 @@
+      <p class="mb-4 text-muted-foreground">No templates yet.</p>
+      <Button @click="openDialog(DialogID.CreateTemplate)">
+        <MdiPlus class="mr-2" />
+        Create Template
```
Needs translating

Wrr2216 commented:
@tonyaellie You're killing me with these translations 😂

Wrr2216 commented:
OK..... I think I have gotten all of the translations (@tonyaellie please don't look lol) 
I also fixed the labels/location and added a test in there to check the template functionality.

Wrr2216 commented:
> @Wrr2216 please share an OpenCollective username, and we'll get the bounty stuff sorted.
> 
> 
> 
> Thanks.

@logan-miller
